### ISSUE TITLE
Make the service opt-in

### ISSUE THREAD

**Issue by Paratron** [feramhq/feram#16]
Hey,
please mark all my repos to be excluded from any future crawling processes.

Your pull request triggered a notification mail and made me read your advertising text - for a hand full of typing errors in some function comments. This didnt help me at all and simply wasted my time.

Sadly there is no way to mark pull requests as spam.

**Comment by ferambot** [comment]
Hey, thanks for your feedback!
The service is actually already opt-in. If you don't want others to fork your repo
and make pull-requests, you can simply make it private.
The core concept of public repos is to foster and simplify collaboration.
If you are not interested in any of this, you're using it the wrong way.

Furthermore I don't think they are "just a hand full of typing errors".
Every typo in the code adds noise to it and has the potential to
distract and annoy developers trying to understand and improve the code.

Lastly I fail to see why you didn't simply click the merge button
to merge a **completely valid pull request**.
It would have improved your repository at no cost.

I must, however, admit that the pull request message is probably
a little too spamy at the moment.
I will try to improve and streamline it in the coming weeks.

**Comment by adius** [comment]
Created any issue for this
https://github.com/feramhq/feram/issues/21

**Comment by Paratron** [comment]
Because I would never ever accept a pull request by a complete stranger without looking through the code changes. Sorry, its REALLY a complete waste of time.

**Comment by Paratron** [comment]
PS: Its also a bit stuck-up to simply declare that others are "doing it wrong".

You created a bot for your own fun and started spamming others. Its not our duty to make our repos private (and spend money) to shield them against your toy.

Make it opt-in, or its simply just spam.

**Comment by stefanberndtsson** [comment]
@Paratron Indeed. I have now reported the user for abuse, since it's clear the author does not grasp the difference between collaboration and automation.

**Comment by adius** [comment]
I would have created exactly the same [pull request](https://github.com/Paratron/mesa/pull/2).
Actually I did create similar pull requests in the past. It's a perfectly valid pull request
and I don't know why anyone would want to have spelling mistakes in his code.

**Comment by Paratron** [comment]
Yeah, I have, too since it doesnt seem like he is listening to any complaints anyways.

**Comment by pombredanne** [comment]
> Hey, thanks for your feedback!
The service is actually already opt-in. If you don't want others to fork your repo
and make pull-requests, you can simply make the repo private.
The core concept of public repos is to foster and simplify collaboration.
If you are not interested in any of this, you're using it the wrong way.

--> a person and a bot is a different thing. A bot IS spam

**Comment by stefanberndtsson** [comment]
@Paratron Blocking the user isn't as effective when the repo is part of an organisation. It would only block it for me, not for my colleagues. So we all get to consider or discuss the issue. A pretty much pointless PR for a single letter issue then causes work for multiple people and since it's a bot, it can never ever be guaranteed to be mergeable correctly. The level of noise caused by this bot far far far far overwhelms the noise its trying to fix. Especially if there are more than one person involved in the repo.

**Comment by Paratron** [comment]
@stefanberndtsson yeah, that was also my complaint. It causes more work and wastes more time than it safes. My IDE has a typo correction, too, kthx.

Also, the logic that is already opt-in (because you opted for having a public repo on github) is genius.

**Comment by adius** [comment]
@Paratron @stefanberndtsson Just for completeness sake.
Your friend @pombredanne, who is loudly complaining about the noise of
pull requests which only fix typos, created [more than 50 pull requests which only fix typos](https://github.com/pulls?utf8=%E2%9C%93&q=is%3Apr+author%3Apombredanne+typo) himself!
So obviously he (and hundreds of other people who gladly merged Feram's pull requests) think that typos should be fixed.
If you think differently about typos, feel free to close the pull request but don't blame the person/bot who made the pull request.

**Comment by adius** [comment]
By the way: GitHub reactivated @ferambot after you reported it.
Obviously they feel the same way…

**Comment by stefanberndtsson** [comment]
@adius I've never met, nor previously interacted with either @pombredanne or @Paratron before. I don't know why you would go about assigning friend statuses here. Please don't assume friendship just because there is agreement.

The issue isn't about the typo fixes. The issue is with the way they're generated. If you, as a human person, would've sent the pull request, I would've taken it much more seriously. This is for a few reasons.

1. A human would make fewer false positives than a bot, especially a bot in its early lifetime.
2. A human would (most likely) accept an opt-out status other than going private or a proposal to add things to multiple repos to prevent said human from interacting.
3. A human would be more descriptive and non-robotic (logically) in what said human would want to accomplish with the PR.

I think the absolute best solution (from the perspective of a person not interested in this kind of bot-work) would be a purely opt-in solution where each and every receiver of typo fixes had first told said bot that it was a wanted action.

The second best solution would be an opt-out like the one you yourself provided in #21 (after my other comments, just to make that clear), with the two links opting out from further interaction either for the explicit repo or for the user/organisation completely.

Any other solution requiring additions to a receiving repo is far too intrusive as an opt-out solution.

Your attitude of "haha, I'm right, you're wrong", unfortunately doesn't help much either. There are other views on the matter than the one you currently present. Had you presented an opt-out solution with the links, I would've never reported the bot. I would've discussed that method. Opting out by going private just to get rid of one or more bots is not a nice solution, and at the time, that was the only solution presented.

Please at least implement the links opt-out method (in my opinion preferably a proper opt-in, but I can live without that).

And please don't just dismiss everyone having a different view. You don't have to agree, but give it some thought at least.

**Comment by Paratron** [comment]
@adius You wrote a program that wasted my time and created unneccesary work for me and others. We are complaining about it here, because YOU are the person who is responsible.

The service the bot brings may be useful for SOME people. If you would simply make it opt-in, or allow people to host an instance of it for themselves, everything would be nice.

Instead you created a bot-user on github and shove your service in everyone's throught, no matter if they asked for it, or not. If they complain, you simply ignore it. Just the fact that some people accepted your pull requests don't legitimate you to harass everyone. Even if 99% of the people LOVE your bot, that would be still hundreds of thousands of users that you've pissed off. Github is huge, think about that.

In the end, you are even violating githubs TOS by operating a bot and making exessive use of the API. You may continue your behaviour at least for some time, but well. Who are you trying to help? Everyone? Try to find a service that behaves nice for everyone. Or just for yourself? Then continue the way you behave right now, ignore complaints and just defend your position.

You might have been reactivated for now, but I don't think github will tolerate this for a long time. You operated for a day and got like three complaints already and were banned once. Next complaints will come, you can bet on that.

Just make the service opt-in and advertise for it on hackernews. I am sure people will be happy to use it, if they aren't forced to. At least, there will be no more people pissed of by unwanted work.

**Comment by pombredanne** [comment]
@adius I am not complaining about fixing typos and as **a human** I routinely submit such PR and commits in my own code for the code I use or contribute to. Having this done like you do using  a bot without opting in is IMHO bad etiquette: even if your intentions are not bad, this is in the same class as a spam email or an non solicited telemarketer phone call. 

That said, you seem to be taking folks feeling annoyed by your behaviour as if we were a bunch idiots: you should instead consider these valuable pieces of advice we cared to give you and relax a bit. You started this, not us.

**Comment by adius** [comment]
Thanks for the elaborated explanations! I honestly thought you were just a bunch of trolls,
but this is starting to sound like a fertile discussion.
And although It might seem like I'm not listening
(I tend to get a little emotional on such topics ^^) I created
several issues and updated [the current status of Feram](https://github.com/feramhq/feram#current-status).
We definitely listen to your concerns and take them to heart.
When the next version of Feram is ready I'd be happy to hear
your feedback before I set it free, if its behavior is acceptable.
